src/vm/analysis/type_checker/mod.rs

```diff
@@ -168,6 +168,33 @@ impl FunctionType {
             },
         }
     }
+
+    pub fn check_args_by_allowing_trait_cast(&self, db: &mut AnalysisDatabase, func_args: &[Value]) -> CheckResult<TypeSignature> {
+        let (expected_args, returns) = match self {
+            FunctionType::Fixed(FixedFunction{ args, returns }) => (args, returns),
+            _ => panic!("Unexpected function type")
+        };
+        check_argument_count(expected_args.len(), func_args)?;
+
+        for (expected_arg, arg) in expected_args.iter().zip(func_args.iter()).into_iter() {
+            match (&expected_arg.signature, arg) {
+                (TypeSignature::TraitReferenceType(trait_id), Value::Principal(PrincipalData::Contract(contract))) => {
+                    let contract_to_check = db.load_contract(contract)
+                        .ok_or_else(|| CheckErrors::NoSuchContract(contract.name.to_string()))?;
+                    let trait_definition = db.get_defined_trait(&trait_id.contract_identifier, &trait_id.name).unwrap()
+                        .ok_or(CheckErrors::NoSuchContract(trait_id.contract_identifier.to_string()))?;
+                    contract_to_check.check_trait_compliance(trait_id, &trait_definition)?;
+                },
+                (expected_type, value) => {
+                    let actual_type = TypeSignature::type_of(&value);
+                    if !expected_type.admits_type(&actual_type) {
```

Comment:
This should use `expected_type.admits(&value)` rather than `admits_type`. Right now, they're equivalent in the VM, but if we ever want to improve the `.admits` function to be more efficient for checking values, we'd want this function to use that function.

Reply:
makes sense.  Addressed in ba8e18245cb9f85ca9ebf2d5541134bdfc855cc5

Reply:
hmm, so much for using `backquote`s in commit logs.  That commits message was:
> address review comment and use `admits` rather than `admits_type` for possible better future proofing